src/exo/cursors.py

```diff
@@ -0,0 +1,599 @@
+from .LoopIR import LoopIR
+
+
+class InvalidCursorError(Exception):
+    pass
+
+
+class ForwardingPolicy(Enum):
+    # Invalidate any cursor that could be forwarded reasonably in more than one
+    # way.
+    PreferInvalidation = auto()
+    # When forwarding insertions, prefer to forward the insertion gap to the gap
+    # BEFORE the inserted block, rather than invalidating it.
+    AnchorPre = auto()
+    # When forwarding insertions, prefer to forward the insertion gap to the gap
+    # AFTER the inserted block, rather than invalidating it.
+    AnchorPost = auto()
+
+
+def _is_sub_range(a: range, b: range):
+    """
+    Returns true if `a` is a STRICT sub-range of `b`.
+    Only applies to step-1 ranges.
+    >>> _is_sub_range(range(1, 4), range(1, 4))
+    False
+    >>> _is_sub_range(range(0, 3), range(3, 6))
+    False
+    >>> _is_sub_range(range(2, 4), range(2, 5))
+    True
+    >>> _is_sub_range(range(2, 4), range(1, 4))
+    True
+    >>> _is_sub_range(range(2, 4), range(1, 5))
+    True
+    >>> _is_sub_range(range(0, 4), range(1, 4))
+    False
+    """
+    assert a.step == b.step and a.step in (1, None)
+    return (a.start >= b.start) and (a.stop <= b.stop) and a != b
+
+
+def _overlaps_one_side(a: range, b: range):
+    """
+    Returns True if `a` overlaps `b` on exactly one side, without containing `b`.
+    Only applies to step-1 ranges.
+    >>> _overlaps_one_side(range(0, 4), range(4, 8))  # fully to left
+    False
+    >>> _overlaps_one_side(range(0, 5), range(4, 8))  # rightmost overlaps leftmost
+    True
+    >>> _overlaps_one_side(range(0, 7), range(4, 8))  # almost contains on right side
+    True
+    >>> _overlaps_one_side(range(0, 8), range(4, 8))  # contains on right side
+    False
+    >>> _overlaps_one_side(range(4, 7), range(4, 8))  # contained, left-aligned
+    True
+    >>> _overlaps_one_side(range(4, 8), range(4, 8))  # equal
+    False
+    >>> _overlaps_one_side(range(5, 8), range(4, 8))  # contained, right-aligned
+    False
+    >>> _overlaps_one_side(range(4, 12), range(4, 8))  # contains on left side
+    False
+    >>> _overlaps_one_side(range(5, 12), range(4, 8))  # almost contains on left side
+    True
+    >>> _overlaps_one_side(range(7, 12), range(4, 8))  # leftmost overlaps rightmost
+    True
+    >>> _overlaps_one_side(range(8, 12), range(4, 8))  # fully to right
+    False
+    """
+    assert a.step == b.step and a.step in (1, None)
+    return (a.start < b.start < a.stop < b.stop or
+            a.start == b.start < a.stop < b.stop or
+            b.start < a.start < b.stop < a.stop)
+
+
+@dataclass
+class Cursor(ABC):
+    _proc: ReferenceType[ProcedureBase]
+
+    # ------------------------------------------------------------------------ #
+    # Static constructors
+    # ------------------------------------------------------------------------ #
+
+    @staticmethod
+    def root(proc: ProcedureBase):
+        return Node(weakref.ref(proc), [])
```

Comment:
@alexreinking @gilbo I still think this function is slightly confusing, because "cursors" is an intuitive concept to users because it directly corresponds to a familiar cursor/selection concept in the text editor. "Having a root of the AST tree" is a real motivation for compiler writers, but for users what is the corresponding intuitive concept? proc? But cursors is not a proc. I'm starting to think what Gilbert was suggesting in the meeting about making root a selection of all statements makes sense.

Reply:
The corresponding intuitive concept is the function AST. I think it makes absolutely no sense at all to restrict cursors from being able to point there. Take for example the following procedure:

```
def example(n: size, a: R[n]):
  assert n > 3
  a[0] = 1
  a[1] = 2
  a[2] = 3
  a[n-1] = 0
```

From an internal data structures perspective, preventing Cursors from being able to point at `n > 3` is unjustifiable to me, period. Doing so would strip cursors of their ability to find and replace all instances of `n` in the whole procedure. To implement such a directive, it would need to loop over each top-level statement and do the replacement separately, then re-write the signature and assertions using... bespoke code? A bespoke forwarding function that's unique to it? Why?

Scaled over an entire scheduling API, this is nonsense. It will lead to a huge amount of redundant work.

I can't justify it from an API perspective either. Imagine you have the above procedure in a text editor and your cursor is pointed at `a[0] = 1`. "What do you mean I can't go up?"

Reply:
I think having a function AST makes sense, but that should have its own class and methods and should be different from the class `Node`. Also I think the method name should be `func()` instead of `root()`, because `root()` is ambiguous.

Regarding pointing cursors to asserts, it is very unintuitive for users if they need to understand that "assertion is an annotation to procedures, not a statement". If we have `func()` or `root()`, users will need to do sth like:
```
c = foo.find_cursor("a[0] = 1")
assert = c.func().assert_stmts[0]
```
to get an assert statement, when they expect `c.prev()` would work.

Reply:
> but that should have its own class and methods and should be different from the class Node.

Maybe as a subtype with fewer functions, but they fundamentally function the same and should share as much implementation as possible.

Reply:
> it is very unintuitive for users if they need to understand that "assertion is an annotation to procedures, not a statement".

I think the fact that assertions _aren't_ statements in unintuitive :shrug:

Reply:
But regardless, they could write `foo.find_cursor("assert e")` and the fact it's an assert PAST tells the matching system where to look.